### DISCUSSION TITLE
How to disable email verification

### DISCUSSION THREAD

**Discussion by agnelvishal** [mastodon/mastodon#27919] [General]
I am working on an alternate authentication system using [web3js](https://web3js.readthedocs.io/en/v1.2.1/) and [Metamask](https://metamask.io)

How do I disable email verification? Which place contains the code?

**Comment by NimaBoscarino** [comment]
I think you'll find the actual mailer code in `app/mailers/user_mailer.rb`. The User model uses Devise, which gives us `:confirmable` for the user, which triggers the verification email. You could experiment with removing `:confirmable`? [This page might help you out.](https://github.com/plataformatec/devise/wiki/How-To:-Add-:confirmable-to-Users)

**Comment by agnelvishal** [comment]
I am new to ruby.
I couldn't find the word `:confirmable ` in https://github.com/tootsuite/mastodon/blob/master/app/mailers/user_mailer.rb

**Comment by NimaBoscarino** [comment]
You'll find it here: `app/models/user.rb`

I definitely recommend taking a look through the [Getting Started with Rails](https://guides.rubyonrails.org/getting_started.html) tutorial! Ruby is a great language, but Rails can be a little intimidating because it has a lot of "magic". The Rails guides are fantastic, however, so I bet you'll feel ready to go in no time.

**Comment by agnelvishal** [comment]
Thanks so much for the advice. I have heard Ruby is a dying language but will definitely try to read the docs.

**Comment by Eliav2** [comment]
what exactly you changed in `app/models/user.rb` in order to disable email verification? there is no option to disable it via management panel in mastodon or via env variable?

**Comment by VityaSchel** [comment]
I was experimenting with this and I think the easiest way is to write confirmed_at: Time.now.utc to db while signing user up and show something like "you was successfully signed up" instead of confirm your email message after filled form. I asked chat gpt about mastodon code and it told me to seek in /db schemas for anything related to postgre operations. I will post solution here if I manage to find one

**Comment by VityaSchel** [comment]
i spent 8 hours on this and found how to auto-confirm all new users
1. open `app/models/user.rb`
2. add to line 123 after `before_create :set_approved` and before `after_commit :send_pending_devise_notifications` this line:
```ruby
before_create :auto_confirm
```
3. on line 454 after block named `def set_approved ... end` and before `def sign_up_from_ip_required_approval? ... end`:
```ruby
def auto_confirm
    self.confirmed_at = Time.now.utc
  end
```

**Comment by beimnetzewdu** [comment]
@VityaSchel hello, hope you are doing well. What you suggested is awesome but I was confronted with the following error can I just remove it or is there any other way.

  .flash-message.warning
    = t('auth.status.confirming')
    = link_to t('auth.didnt_get_confirmation'), new_user_confirmation_path
- elsif !user.approved?
  .flash-message.warning
    = t('auth.status.pending')
